File: keyplayer.py
import numpy as np
import bisect
import random
import sys

# pick k numbers less than n
def get_random_set(n, k):
    T = range(n)
    S = []
    for i in range(k):
        j = random.randint(0, len(T)-1)
        bisect.insort(S, T.pop(j)) # insert into sorted list
    return S, T


# in sorted list, take out x and add y
def switch(sl, x, y):
    sl.remove(x)
    bisect.insort(sl, y)


def matsquare(M):
    N = np.asarray(M)
    n, m = M.shape
    for i in range(n):
        for j in range(n):
            for k in range(n):
                N[i,j] = min(N[i,j], M[i, k] + M[k, j])
    return N


def graph_distance(A):
    D = np.asarray(A, dtype=np.double)
    D[D==0] = np.inf
    while True:
        D_ = matsquare(D)
        if (D==D_).all():
            return D
        D = D_

def reachability(A):
    D = graph_distance(A)
    R = np.asarray([1 if d > 0 and d < np.inf else 0 for d in D.flatten()])
    R.shape = D.shape
    return R


# KPP-Neg metric:

# number of disconnected pairs (Borgatti (3))
# minimize sum(i>j) R_ij
# Equation (4) should be more tractable in the optimization context (although lose out on elegance of equation)
def metric_3(A):
    R = reachability(A)
    s = np.sum(np.tril(R, k=-1))
    return s

# Borghatti (9)
# maximize D_F = 1 - 2 * (sum(i>j) (1 / d_ij) / (n * n-1)
# equivalent to minimizing just the sum
def metric_9(A):
    D = graph_distance(A)
    return np.sum(np.tril(1/D, k=-1))


def greedy_optimize(A, k, metric, tolerance):
    S, T = get_random_set(len(A), k) # sorted. S union T = {1 ... |A|}
    
    B = A[T, :][:, T] # A without col i and row i if i in S
    fit = metric(B)

    print "Fit: %s" % (fit)
    while True:
        Dfit = np.inf
        pair = None
        for u in S:
            for v in T:
                T_ = list(T)
                switch(T_, v, u)
                B_ = A[T_, :][:, T_]
                fit_ = metric(B_)
                d = fit - fit_
                if (d >= 0) and (d < Dfit):
                    Dfit = d
                    pair = u,v
        if (Dfit < tolerance):
            break
        u, v = pair
        switch(S, u, v)
        fit = fit - d

    print "New fit: %s" % (fit)
    return S

# random graph adjacency matrix
def rgraph(n, p):
    m = np.zeros(n*n)
    for i in range(n*n):
        if (np.random.rand() > p):
            m[i]=1
    m.shape = (n,n)
    return m

def main(argv):
    n = int(argv[1]) # number of nodes in graph
    p = float(argv[2]) # probability two nodes are connected
    k = int(argv[3]) # number of nodes to find
    tol = float(argv[4]) # tolerance to stop optimize algorithm at
    t = int(argv[5]) # times to repeat the optimize algorithm
    m = int(argv[6]) # metric (3 or 9)
   
    if (m == 3):
        metric = metric_3
    elif (m == 9):
        metric = metric_9
    else:
        print "Invalid metric! Use 3 or 9 only."

    G = rgraph(n, 0.5)

    for i in range(t):
        S = greedy_optimize(G, k, metric, tol)
        print "nodes: %s" % S
        print

if __name__ == "__main__":
    main(sys.argv)

File: keyplayer.R
library(statnet)

# pick k numbers less than n. return S, T, where |S|=k and |T|=n-k, S union T = {1,2,..,n} 
# no longer used, switched to using a mask (see get_random_mask)

get_random_set <- function(G, k) {
    # input G as the graph
    n <- network.size(G)
    range <- 1:n
    samp <- sample(n, k)
    s <- range[range %in% samp]
    t <- range[! range %in% samp]
    return (list(s,t));
}


graph_distance <- function(x) { 
    return(geodist(x)$gdist);
}

reachability = function(A) {
    D = graph_distance(A)
    return ((D > 0) & (D < Inf));
}

# trim array based on True, False mask (see get_random_mask above)
trimmed_array = function(A, s) {
    B <- A[s,][,s]
    return(B);
}

# A *metric* is a function that takes an adjancancy matrix G an indicator vector s, where:
#   s = [s1, s2, s3, .. sn] where s_i = TRUE iff node i in S
# and returns a number between 0 and 1.

# KPP-Neg metric:

# number of disconnected pairs (Borgatti (3))
# maximize D =  1 - 2 * sum(i>j) R_ij / (n * (n-1))
# Equation (4) should be more tractable in the optimization context (although lose out on elegance of equation)
metric_3 <- function(G, s) {
    A <- trimmed_array(G, !s)
    n <- length(A[1,])
    R <- reachability(A);
    S <- sum(R[lower.tri(R)]);
    return(1 - 2*S/(n * (n-1)));
}

# Borghatti (9)
# maximize D_F = 1 - 2 * (sum(i>j) (1 / d_ij) / (n * (n-1))
metric_9 <- function(G, s) {
    A <- trimmed_array(G, !s)
    n <- length(A[1,])
    D <- graph_distance(A);
    Dr <- 1/D
    S <- sum(Dr[lower.tri(D)]);
    return(1 - 2*S/(n * (n-1)));
}

# KPP-Pos metric:

# Borghatti (14)
# d_Kj = min { d(i, j) where i in K, j in G\S }
# D_R = sum(j) (1/d_Kj) / n
metric_14 <- function(G, s) {
    n <- length(G[1,])
    D <- graph_distance(G) # future optimization: only do the distance calculation ONE time.

    H <- apply(D[!s, s], 1, min); # 0 < d <= Inf 
    return (sum(1/H)/n)
}


# return something like [TRUE, FALSE, FALSE, TRUE, ...] where there are n total values and k FALSEs
get_random_mask = function(n, k) {
    s <- sample(n, k);
    s <- 1:n %in% s;
    return (s);
}


greedy_optimize = function(A, k, metric, tolerance) {
    n <- length(A[1,])

    s <- get_random_mask(n, k); # random index matrix to start with.
                                # our "target" nodes are {i | s[i] is TRUE}


    fit <- metric(A, s)

    print(paste("Fit: ", fit))
    i <- 0
    while (TRUE) {
        i <- i + 1
        Dfit = 0
        pair = NULL
        for (u in which(s)) {
            for (v in which(!s)) {
                s_ <- s # clone
                s_[v] = TRUE;
                s_[u] = FALSE;
                
                fit_ = metric(A, s_)
                d = fit_ - fit
                if ((d >= 0) && (d > Dfit)) {
                    Dfit = d
                    pair = list(u,v)
                }
            }
        }
        if (Dfit < tolerance)
            break
        print(paste("Iteration",i,":", fit, "=>",fit+Dfit))
        u <- pair[[1]]
        v <- pair[[2]]
        s[v] = TRUE
        s[u] = FALSE
        fit = fit + Dfit
    }

    print(paste("New fit (", i, " iterations): ", fit));
    return(which(s));
}


main = function(argv) {
    ## n = as.numeric(argv[1]) # number of nodes in graph
    ## p = as.numeric(argv[2]) # probability two nodes are connected
    G = as.network(argv[[1]])
    k = as.numeric(argv[[2]]) # number of nodes to find
    tol = as.numeric(argv[[3]]) # tolerance to stop optimize algorithm at
    t = as.numeric(argv[[4]]) # times to repeat the optimize algorithm
    m = as.numeric(argv[[5]]) # metric (3 or 9)
   
    if (m == 3)
        metric <- metric_3
    else if (m == 9)
        metric <- metric_9
    else if (m == 14)
        metric <- metric_14
    else {
        print("Invalid metric! Use 3, 9, or 14 only.")
        q()
    }

    A <- as.matrix(G)
    for (i in 1:t) {
        S <- greedy_optimize(A, k, metric, tol)
        print(sprintf("nodes: %s", paste(S, collapse=",")))
        cat('\n')
    }
}


argv <- commandArgs(trailingOnly = TRUE);
if (length(argv) > 1)
    main(argv)

